runtime/gc_vlhgc/VLHGCAccessBarrier.cpp

```diff
@@ -435,61 +435,96 @@ MM_VLHGCAccessBarrier::jniReleasePrimitiveArrayCritical(J9VMThread* vmThread, ja
 	VM_VMAccess::inlineExitVMToJNI(vmThread);
 }
 
+void
+MM_VLHGCAccessBarrier::copyStringCritical(J9VMThread *vmThread, GC_ArrayObjectModel *indexableObjectModel,
+	J9InternalVMFunctions *functions, jchar **data, J9JavaVM *javaVM,
+	J9IndexableObject *valueObject, J9Object *stringObject, jboolean *isCopy, bool isCompressed)
+{
+	jint length = J9VMJAVALANGSTRING_LENGTH(vmThread, stringObject);
+	UDATA sizeInBytes = length * sizeof(jchar);
+	*data = (jchar*)functions->jniArrayAllocateMemoryFromThread(vmThread, sizeInBytes);
+	if (NULL == *data) {
+		functions->setNativeOutOfMemoryError(vmThread, 0, 0);	// better error message here?
+	} else {
+		if (isCompressed) {
+			jint i;
+			for (i = 0; i < length; i++) {
+				*data[i] = (jchar)J9JAVAARRAYOFBYTE_LOAD(vmThread, (j9object_t)valueObject, i) & (jchar)0xFF;
+			}
+		} else {
+			if (J9_ARE_ANY_BITS_SET(javaVM->runtimeFlags, J9_RUNTIME_STRING_BYTE_ARRAY)) {
+				// This API determines the stride based on the type of valueObject so in the [B case we must passin the length in bytes
+				indexableObjectModel->memcpyFromArray(*data, valueObject, 0, (I_32)sizeInBytes);
+			} else {
+				indexableObjectModel->memcpyFromArray(*data, valueObject, 0, length);
+			}
+		}
+		if (NULL != isCopy) {
+			*isCopy = JNI_TRUE;
+		}
+		vmThread->jniCriticalCopyCount += 1;
+	}
+}
+
 const jchar*
 MM_VLHGCAccessBarrier::jniGetStringCritical(J9VMThread* vmThread, jstring str, jboolean *isCopy)
 {
 	jchar *data = NULL;
 	J9JavaVM *javaVM = vmThread->javaVM;
 	J9InternalVMFunctions *functions = javaVM->internalVMFunctions;
 	VM_VMAccess::inlineEnterVMFromJNI(vmThread);
+	GC_ArrayObjectModel *indexableObjectModel = &_extensions->indexableObjectModel;
 
 	J9Object *stringObject = (J9Object*)J9_JNI_UNWRAP_REFERENCE(str);
 	J9IndexableObject *valueObject = (J9IndexableObject*)J9VMJAVALANGSTRING_VALUE(vmThread, stringObject);
-	bool shouldCopy = false;
 	bool isCompressed = false;
 
 	/* If the string bytes are in compressed UNICODE, then we need to copy to decompress */	
 	if (IS_STRING_COMPRESSED(vmThread, stringObject)) {
 		isCompressed = true;
-		shouldCopy = true;
 	}
```

Comment:
not that shouldCopy  is gone, you can make this one liner:

bool isCompressed = IS_STRING_COMPRESSED(vmThread, stringObject);